# COR1009-5 Quiz 1

- Instructor: 서용덕
- 5 Points Each.

- You must not delete any cell. They should exist as provided.
- You may create cells to do your own computation, or to verify the solution of yours.

# 1. What is your name and student id number? 

You must use python's `print()` function to print them out.

In [1]:
# your code here
print("서용덕, 8803190")

# 2. Use `for`-loop to calculate the sum of all the numbers in the arrary `A`.
- No numpy method functions are allowed. You won't achieve any point.

In [2]:
# do not modify.
import numpy as np
rng = np.random.default_rng(2021) 
array_size = 1000
A = rng.integers(0, 1000, size=array_size)

In [3]:
# your code must come here
sumall = 0
for a in A:
    sumall += a
print('sum of A is ', sumall)

sum of A is  490197


# 3. Two functions to compute the maximum and the minimum, respectively, of the input array `B` using `for`-loop.

In [9]:
# do not modify.
import numpy as np
rng = np.random.default_rng(2022) 
array_size = 1000
B = rng.integers(0, 1000, size=array_size)

In [10]:
# your code here to define functions that computes min and max, respectively.

def my_max(A):
    maxval = A[0]
    for a in A:
        if maxval < a:
            maxval = a
    return maxval

def my_min(A):
    minval = A[0]
    for a in A:
        if minval > a:
            minval = a
    return minval 

In [11]:
# just execute the cell
# call the function to compute max and min of the array, print out the result of function calls
maxvalue = my_max(B)
minvalue = my_min(B)
print(f"min = {minvalue}, max = {maxvalue}.")

min = 0, max = 999.


# 4. Find the sample variance `Var` of the array `A` using for-loop.

The fomula to compute sample variance is as follows:
$$
    Var = \frac{1}{N-1} \sum_{i=1}^N \big( A[i] - m \big)^2
$$
where $m$ is the sample mean computed by:
$$
    m = \frac{1}{N} \sum_{i=1}^N A[i]
$$

In [26]:
# do not modify.
import numpy as np
rng = np.random.default_rng(2023) 
array_size = 1000
A = rng.integers(0, 1000, size=array_size)

In [27]:
# define your function to compute the variance.
def mean(A):
    sumall = 0
    for a in A:
        sumall += a
    return sumall / float(len(A))

def variance(A):
    var = 0 
    m = mean(A)
    for a in A:
        var += (a-m)**2
    var /= float(len(A) - 1)
    return var

In [29]:
# call the function to compute the variance of data in array A
variance(A)

83861.53721621615

# 5. Print the following shape for n = 4.

E.g. n = 3
<pre>
1 3 5 5 3 1
3 5     5 3
5         5
5         5
3 5     5 3
1 3 5 5 3 1
</pre>

In [15]:
# your code comes here.


# 6. Greatest Position Distance Between Matching Array Values

The goal of this problem is to return the greatest distance of index positions between matching number values in an array or 0 if there are no matching values.

Example: In an array with the values [0, 2, 1, 2, 4, 1] the greatest index distance is between the matching '1' values at index 2 and 5. Executing greatestDistance against this array would return 3. (i.e. 5 - 2)

Here's an example:

```
greatest_distance([0, 2, 1, 2, 4, 1])  --> 3
```

In [61]:
# provide  your solution
def greatest_distance(array):
    maxdist = 0
    for i in range(len(array)):
        for j in range(i+1, len(array)):
            if array[i] == array[j]:
                dist = j - i
                if maxdist < dist:
                    maxdist = dist 
    return maxdist

In [62]:
# a utility function. Dont modify. Just run.
def assert_equals(a, b):
    return a == b

In [63]:
# just execute this cell to check whether your function works or not
print(assert_equals(greatest_distance([9, 7, 1, 2, 3, 7, 0, -1, -2]), 4))
print(assert_equals(greatest_distance([0,7,0,2,3,7,0,-1,-2]), 6))

True
True


# 7. Find common items in two lists

1. find those common items from `lista` and `listb`. You must develop your own function with for-loop iteration. 
2. print out the number of common items.
3. find the maximum among the common items. You must use your own `my_max()` function.
4. find the minimum among the common items. You must use your own `my_min()` function.
5. Do not print all of the common items, or you will get negative reward.

In [32]:
# data generation. do not modify.
import numpy as np
rng = np.random.default_rng(2027)

lista = list(rng.integers(0,100,size=200))
listb = list(rng.integers(0,100,size=200))

In [33]:
# your code here
def common_items(la, lb):
    com = []
    for a in la:
        for b in lb:
            if a == b:
                if a not in com:
                    com.append(a)
    return com 

In [35]:
# just run this cell. Do not edit.
list_common = common_items(lista, listb)
print(list_common)

[63, 0, 14, 38, 8, 47, 49, 91, 44, 68, 70, 65, 29, 12, 62, 50, 11, 73, 72, 2, 85, 96, 87, 74, 36, 53, 18, 21, 58, 71, 86, 90, 4, 10, 7, 31, 79, 84, 13, 83, 64, 82, 92, 23, 20, 30, 33, 55, 9, 22, 24, 34, 97, 69, 25, 88, 77, 41, 66, 94, 46, 16, 15, 98, 80, 81, 51, 78, 26, 76, 67, 37, 52, 5, 1, 59, 93, 6]


In [36]:
# find min and max, and print the results.
my_min(list_common), my_max(list_common)

(0, 98)

# 8. Opposites Attract

Timmy & Sarah think they are in love, but around where they live, they will only know once they pick a flower each. If one of the flowers has an even number of petals and the other has an odd number of petals it means they are in love.

Write a function that will take the number of petals of each flower and return true if they are in love and false if they aren't.


In [37]:
# develop your function here
def love(petals1, petals2):
    if petals1 % 2 == 1 and petals2 % 2 == 0:
        return True 
    elif petals2 % 2 == 1 and petals1 % 2 == 0:
        return True 
    else:
        return False 

In [38]:
# just run the cell
print(love(1, 4))

True


In [39]:
# just run the cell
print(love(1101, 9984))

True


# 9. Convert a number to a string!

We need a function that can transform a number into a string.

What ways of achieving this do you know?

123 --> "123"

999 --> "999"

In [42]:
def num2str(num):
    return str(num)

In [44]:
# simply run the cell
a = num2str(6724)
print(a, type(a))

6724 <class 'str'>


In [60]:
# simply run the cell
a = num2str(3141595357989)
print(a, type(a))

3141595357989 <class 'str'>


# 10. Unique items

Make a function that finds all the unique items, which may be fulfilled with the numpy function `np.unique()`.
- Use for-loop.
- You cannot use `np.unique()`, of course.

In [45]:
# For example, this cell show the input and output of `np.unique()`
values = [1, 2, 2, 1, 3, 3, 1, 2]
uniqs = np.unique(values)
print(uniqs)

[1 2 3]


In [48]:
# You are asked to make the following function.
def my_unique(array):
    uniq = []
    for a in array:
        if a not in uniq:
            uniq.append(a)
    return uniq

In [49]:
# do not modify. Just execute the cell.
uniq_values = my_unique(values)
print(uniq_values)

[1, 2, 3]


In [51]:
# In the next cell (not this cell), 
# find min and max of the uniq_values of the following array, 
# calcuate sum of the values, and print out
import numpy as np
rng = np.random.default_rng(2027)

array = list(rng.integers(0,100,size=200))

In [52]:
# ---- your code starts from here ----
coms = my_unique (array)
mmsum = my_max(coms) + my_min(coms)
print(mmsum)

99


# 11 (Extra 10) Number of people in the Bus

There is a bus moving in the city, and it takes and drop some people in each bus stop.

You are provided with a list (or array) of integer pairs. Elements of each pair represent number of people get into bus (The first item) and number of people get off the bus (The second item) in a bus stop.

Your task is to return number of people who are still in the bus after the last bus station (after the last array). Even though it is the last bus stop, the bus is not empty and some people are still in the bus, and they are probably sleeping there :D

Take a look on the test cases.

Please keep in mind that the test cases ensure that the number of people in the bus is always >= 0. So the return integer can't be negative.

The second value in the first integer array is 0, since the bus is empty in the first bus stop.

In [55]:
def number(bus_stops):
    ins = 0
    outs = 0
    for ios in bus_stops:
        i, o = ios 
        ins += i
        outs += o 
    finale = ins - outs 
    return finale 
    # Good Luck

If your function works well, you will see the following cells print Trues.

In [56]:
# run the cell
print(number([[10,0],[3,5],[5,8]]) == 5)

True


In [57]:
# run the cell
print(number([[3,0],[9,1],[4,10],[12,2],[6,1],[7,10]]) == 17)

True


In [58]:
# run the cell
print(number([[3,0],[9,1],[4,8],[12,2],[6,1],[7,8]]) == 21)

True


In [59]:
# run the cell
stops = [[89, 0], [14, 25], [96, 77], [86, 52], [41, 76], [35, 21], [43, 53], [64, 21], [7, 16], [9, 129], [73, 8], [64, 67], [2, 63], [34, 0], [33, 11], [15, 59], [25, 2], [35, 11], [48, 1], [23, 114], [33, 20], [51, 24], [89, 28], [79, 55], [0, 121], [74, 3], [8, 65], [28, 23], [31, 41], [14, 14], [64, 34], [82, 52], [77, 81], [26, 95], [16, 3], [98, 0], [54, 117], [66, 29], [19, 67], [64, 7], [75, 123], [34, 43], [92, 52], [26, 38], [11, 76], [7, 2], [24, 5], [20, 39], [52, 9], [84, 36], [56, 34], [54, 59], [63, 131], [99, 30], [18, 140], [20, 15], [7, 7], [98, 7], [78, 101], [26, 6], [88, 108], [45, 23], [70, 54], [30, 40], [41, 70], [97, 2], [2, 70], [55, 43], [48, 43], [46, 125], [8, 30], [55, 26], [90, 64], [52, 72], [81, 0], [22, 42], [47, 71], [4, 52], [59, 18], [61, 2], [17, 106], [90, 56], [8, 21], [67, 58], [69, 4], [8, 130], [82, 43], [55, 31], [94, 104], [20, 67], [65, 16], [17, 41], [62, 18], [97, 112], [97, 20], [86, 63], [81, 60], [46, 26], [61, 156], [88, 138], [15, 6]]

print(number(stops) == 110)

True


End.